content/blaze.md

```diff
@@ -0,0 +1,543 @@
+});
+
+Template.todosItem.events({
+  'focus input[type=text]'() {
+    this.onEdit(true);
+  }
+});
+```
+
+<h3 id="onrendered-for-libs">Use `onRendered()` to callout to 3rd party libraries</h3>
+As we mentioned above, the `onRendered()` callback is typically the right spot to call out to third party libraries that expect a pre-rendered DOM (such as jQuery plugins). The `onRendered()` callback is triggered *once* after the template is rendered and attached to the DOM for the first time.
+
+Occasionally, you may need to wait for data to become ready before it's time to attach the plugin (although typically it's a better idea to use a sub-template in this use case). To do so, you can setup an `autorun` in the `onRendered()` callback. For instance, in the `listShowPage` template, we want to wait until the subscription for the list is ready (i.e. the todos have rendered) before we hide the launch screen:
+
+```
+Template.listsShowPage.onRendered(function() {
+  this.autorun(() => {
+    if (this.subscriptionsReady()) {
+      // Handle for launch screen defined in app-body.js
+      AppLaunchScreen.listRender.release();
+    }
+  });
+});
+```
+
+<h2 id="smart-components">Writing smart components with Blaze</h2>
+If your component needs to access state outside of its data context---for instance, data from the server via subscriptions or the the contents of client-side store, then you should be careful how you do that accessing. As discussed in the [data loading article](data-loading) you should be careful and considered in how use such smart components.
+
+To begin with, all of the rules of thumb about reusable components apply to smart components. In addition:
+
+<h3 id="subscribing">Subscribe from `onCreated`</h3>
+You should subscribe to publications on the server from an `onCreated` callback (within an `autorun` block if you are reliant on reactively changing arguments). In the Todos example app, in the `listsShowPage` template we subscribe to the `list/todos` publication based on the current `_id` FlowRouter param:
+
+```js
+Template.listsShowPage.onCreated(function() {
+  this.state = new ReactiveDict();
+  this.autorun(() => {
+    const listId = FlowRouter.getParam('_id');
+    this.state.set({listId});
+    this.subscribe('list/todos', listId);
+  });
+});
+```
+
+By using `this.subscribe()` (as opposed to `Meteor.subscribe`), the subscription state automatically gets amalagamated into the template instance's subscription readiness reactive state variable, which can be used both from within templates (via the `subscriptionsReady` helper) or within helpers (via `Template.subscriptionsReady()`).
+
+Notice as well in this case that we access the global client-side state store `FlowRouter` in the `onCreated()` callback also, storing its value locally in a local `state` dictionary.
+
+XXX: should instead we just read from flow router a second time here: https://github.com/meteor/todos/blob/master/packages/lists-show/lists-show-page.js#L28 ??
```

Comment:
I guess the logic is, we want to do all "smart" things inside `onCreated` rather than inside helpers? One more way to achieve that would be to define a function inside `onCreated` like:

```js
this.getListId = () => FlowRouter.getParam('_id')

this.autorun(() => {
  this.subscribe('list/todos', this.getListId());
});
```

And then use that function inside the helper. That way if you need to get the listId from somewhere else, you don't need to change the helper?